bpf/command.bpf.c

```diff
@@ -0,0 +1,162 @@
+    u64 ppid;
+    char comm[TASK_COMM_LEN];
+    enum event_type type;
+    char argv[ARGSIZE];
+    int retval;
+    u64 cgroup;
+};
+
+BPF_RING_BUF(execve_events, 4096*8);
+
+static int __submit_arg(void *ptr, struct data_t *data)
+{
+    bpf_probe_read(data->argv, sizeof(data->argv), ptr);
+    bpf_ringbuf_output(&execve_events, data, sizeof(struct data_t), 0);
+    return 1;
+}
+
+static int submit_arg(void *ptr, struct data_t *data)
+{
+    const char *argp = 0;
+    bpf_probe_read(&argp, sizeof(argp), ptr);
+    if (argp) {
+        return __submit_arg((void *)(argp), data);
+    }
+    return 0;
+}
+
+static pid_t get_ppid(struct task_struct *task) {
+    // Some kernels, like Ubuntu 4.13.0-generic, return 0
+    // as the real_parent->tgid.
+    // We use the getPpid function as a fallback in those cases.
+    // See https://github.com/iovisor/bcc/issues/1883.
+    return BPF_CORE_READ(task, real_parent, tgid);
+}
+
+static int enter_execve(const char *filename,
+                const char *const *argv,
+                const char *const *envp)
+{
+    // create data here and pass to submit_arg to save stack space (#555)
+    struct data_t data = {};
+    struct task_struct *task;
+
+    data.pid = bpf_get_current_pid_tgid() >> 32;
+    data.cgroup = bpf_get_current_cgroup_id();
+
+    task = (struct task_struct *)bpf_get_current_task();
+    // Some kernels, like Ubuntu 4.13.0-generic, return 0
+    // as the real_parent->tgid.
+    // We use the getPpid function as a fallback in those cases.
+    // See https://github.com/iovisor/bcc/issues/1883.
+    data.ppid = get_ppid(task);
+
+    bpf_get_current_comm(&data.comm, sizeof(data.comm));
+    data.type = EVENT_ARG;
+
+    __submit_arg((void *)filename, &data);
+
+    // skip first arg, as we submitted filename
+    #pragma unroll
+    for (int i = 1; i < 20; i++) {
+        if (submit_arg((void *)&argv[i], &data) == 0)
+             goto out;
+    }
+
+    // handle truncated argument list
+    char ellipsis[] = "...";
+    __submit_arg((void *)ellipsis, &data);
+out:
+    return 0;
+}
+
+static int exit_execve(int ret)
+{
+    struct data_t data = {};
+    struct task_struct *task;
+
+    data.pid = bpf_get_current_pid_tgid() >> 32;
+    data.cgroup = bpf_get_current_cgroup_id();
+
+    task = (struct task_struct *)bpf_get_current_task();
+    // Some kernels, like Ubuntu 4.13.0-generic, return 0
+    // as the real_parent->tgid.
+    // We use the getPpid function as a fallback in those cases.
+    // See https://github.com/iovisor/bcc/issues/1883.
+    data.ppid = BPF_CORE_READ(task, real_parent, tgid);
+
+    bpf_get_current_comm(&data.comm, sizeof(data.comm));
+    data.type = EVENT_RET;
+    data.retval = ret;
+    bpf_ringbuf_output(&execve_events, &data, sizeof(data), 0);
+
+    return 0;
+}
+
+SEC("tp/syscalls/sys_execve")
+int tracepoint__syscalls__sys_enter_execve(struct trace_event_raw_sys_enter *tp)
+{
+    const char *filename = (const char *)tp->args[0];
```

Comment:
Dumb question, but here (and elsewhere) we're guaranteed to always be able to index into `args` like this right? No need to do a length check here?

Reply:
I believe so (unless there's a kernel bug). The length of the args will match the number of args to the syscall.